File: hooks/scripts/statusline.ts
// Statusline - Claude Code custom statusline entry point
// Reads JSON from stdin, renders based on config mode (full/compact/minimal/none)

import { stdin } from "node:process";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { getGitInfo } from "./lib/git-info-cache.ts";
import { loadConfig } from "./lib/config-loader.ts";
import { parseTranscript } from "./lib/transcript-parser.ts";
import { collapseHome } from "./lib/statusline-utils.ts";
import { renderFull, renderCompact, renderMinimal, type RenderContext } from "./lib/statusline-renderer.ts";

// Buffer constant matching /context output (22.5% of 200k)
const AUTOCOMPACT_BUFFER = 45000;

/** Read all stdin as string */
async function readStdin(): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: string[] = [];
    stdin.setEncoding("utf8");
    stdin.on("data", (chunk: string) => chunks.push(chunk));
    stdin.on("end", () => resolve(chunks.join("")));
    stdin.on("error", reject);
  });
}

async function main(): Promise<void> {
  try {
    const input = await readStdin();
    if (!input.trim()) {
      console.error("No input provided");
      process.exit(1);
    }

    const data = JSON.parse(input);

    // Extract basic info
    const rawDir = data.workspace?.current_dir || data.cwd || process.cwd();
    const currentDir = collapseHome(rawDir);
    const modelName = data.model?.display_name || "Claude";

    // Git info (cached 30s)
    const gitInfo = getGitInfo(rawDir);

    // Context window calculation with autocompact buffer
    const usage = data.context_window?.current_usage || {};
    const contextSize = data.context_window?.context_window_size || 0;
    let contextPercent = 0;
    let totalTokens = 0;

    if (contextSize > 0 && contextSize > AUTOCOMPACT_BUFFER) {
      totalTokens = (usage.input_tokens ?? 0) +
        (usage.cache_creation_input_tokens ?? 0) +
        (usage.cache_read_input_tokens ?? 0);
      contextPercent = Math.min(100, Math.round(((totalTokens + AUTOCOMPACT_BUFFER) / contextSize) * 100));
    }

    // Write context data to temp file for hooks to read
    const sessionId = data.session_id;
    if (sessionId && contextSize > 0) {
      try {
        const contextDataPath = path.join(os.tmpdir(), `sl-context-${sessionId}.json`);
        fs.writeFileSync(contextDataPath, JSON.stringify({
          percent: contextPercent,
          tokens: totalTokens,
          size: contextSize,
          usage,
          timestamp: Date.now(),
        }));
      } catch { /* silent */ }
    }

    // Parse transcript for tools/agents/todos
    const transcriptPath = data.transcript_path;
    const transcript = transcriptPath
      ? await parseTranscript(transcriptPath)
      : { tools: [], agents: [], todos: [], sessionStart: null };

    // Read usage limits from cache (written by usage-context-awareness hook)
    let sessionText = "";
    let usagePercent: number | null = null;
    try {
      const usageCachePath = path.join(os.tmpdir(), "sl-usage-limits-cache.json");
      if (fs.existsSync(usageCachePath)) {
        const cache = JSON.parse(fs.readFileSync(usageCachePath, "utf8"));
        if (cache.status === "unavailable") {
          sessionText = "N/A";
        } else {
          const fiveHour = cache.data?.five_hour;
          usagePercent = fiveHour?.utilization ?? null;
          const resetAt = fiveHour?.resets_at;
          if (resetAt) {
            const resetTime = new Date(resetAt);
            const remaining = Math.floor(resetTime.getTime() / 1000) - Math.floor(Date.now() / 1000);
            if (remaining > 0 && remaining < 18000) {
              const rh = Math.floor(remaining / 3600);
              const rm = Math.floor((remaining % 3600) / 60);
              sessionText = `${rh}h ${rm}m until reset`;
            }
          }
        }
      }
    } catch { /* silent */ }

    // Lines changed
    const linesAdded = data.cost?.total_lines_added || 0;
    const linesRemoved = data.cost?.total_lines_removed || 0;

    // Build render context
    const ctx: RenderContext = {
      modelName,
      currentDir,
      gitBranch: gitInfo?.branch || "",
      gitUnstaged: gitInfo?.unstaged || 0,
      gitStaged: gitInfo?.staged || 0,
      gitAhead: gitInfo?.ahead || 0,
      gitBehind: gitInfo?.behind || 0,
      contextPercent,
      sessionText,
      usagePercent,
      linesAdded,
      linesRemoved,
      transcript,
    };

    // Load config for statusline mode
    const config = loadConfig({ includeProject: false, includeAssertions: false, includeLocale: false });
    const statuslineMode = config.statusline || "full";

    // Render and output
    let lines: string[];
    switch (statuslineMode) {
      case "none":
        lines = [""];
        break;
      case "minimal":
        lines = renderMinimal(ctx);
        break;
      case "compact":
        lines = renderCompact(ctx);
        break;
      case "full":
      default:
        lines = renderFull(ctx);
        break;
    }

    for (const line of lines) {
      console.log(line);
    }
  } catch {
    // Fallback: minimal dir line on any error
    console.log("📁 " + (process.cwd() || "unknown"));
  }
}

main().catch(() => {
  console.log("📁 error");
  process.exit(1);
});

File: hooks/scripts/cmd/statusline.go
// statusline.go: statusline subcommand — renders Claude Code status display.
// Reads JSON from stdin, builds render context, outputs lines to stdout.
// Falls back to "📁 dir" on any error so Claude Code always gets a valid line.
package cmd

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"os"
	"path/filepath"
	"time"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	"solon-hooks/internal/git"
	"solon-hooks/internal/statusline"
)

// autocompactBuffer matches TS constant: 22.5% of 200k context.
const autocompactBuffer = 45000

var statuslineCmd = &cobra.Command{
	Use:   "statusline",
	Short: "Render Claude Code status line",
	RunE:  runStatusline,
}

func runStatusline(cmd *cobra.Command, args []string) error {
	raw, err := io.ReadAll(os.Stdin)
	if err != nil || len(raw) == 0 {
		printFallback()
		return nil
	}
	if err := renderStatusline(raw); err != nil {
		printFallback()
	}
	return nil
}

// printFallback prints the minimal fallback line when anything goes wrong.
func printFallback() {
	cwd, err := os.Getwd()
	if err != nil {
		cwd = "unknown"
	}
	fmt.Println("📁 " + statusline.CollapseHome(cwd))
}

// renderStatusline parses stdin JSON and renders the full statusline output.
func renderStatusline(raw []byte) error {
	var data map[string]interface{}
	if err := json.Unmarshal(raw, &data); err != nil {
		return err
	}

	// Extract working directory
	rawDir := ""
	if ws, ok := data["workspace"].(map[string]interface{}); ok {
		rawDir, _ = ws["current_dir"].(string)
	}
	if rawDir == "" {
		rawDir, _ = data["cwd"].(string)
	}
	if rawDir == "" {
		rawDir, _ = os.Getwd()
	}
	currentDir := statusline.CollapseHome(rawDir)

	// Model display name
	modelName := "Claude"
	if m, ok := data["model"].(map[string]interface{}); ok {
		if v, ok := m["display_name"].(string); ok && v != "" {
			modelName = v
		}
	}

	// Git info (30s cached)
	gitInfo := git.GetGitInfo(rawDir)

	// Context window calculation
	contextPercent := 0
	totalTokens := 0
	contextSize := 0
	if cw, ok := data["context_window"].(map[string]interface{}); ok {
		if v, ok := jsonInt(cw, "context_window_size"); ok {
			contextSize = v
		}
		if usage, ok := cw["current_usage"].(map[string]interface{}); ok && contextSize > autocompactBuffer {
			inp, _ := jsonInt(usage, "input_tokens")
			cacheCreate, _ := jsonInt(usage, "cache_creation_input_tokens")
			cacheRead, _ := jsonInt(usage, "cache_read_input_tokens")
			totalTokens = inp + cacheCreate + cacheRead
			contextPercent = int(math.Min(100, math.Round(
				float64(totalTokens+autocompactBuffer)/float64(contextSize)*100,
			)))
		}
	}

	// Write context data to temp file for other hooks
	sessionID, _ := data["session_id"].(string)
	if sessionID != "" && contextSize > 0 {
		writeContextCache(sessionID, contextPercent, totalTokens, contextSize, data)
	}

	// Parse transcript
	transcriptPath, _ := data["transcript_path"].(string)
	transcript := statusline.ParseTranscript(transcriptPath)

	// Read usage limits cache
	sessionText, usagePercent := readUsageLimitsCache()

	// Lines changed
	linesAdded, linesRemoved := 0, 0
	if cost, ok := data["cost"].(map[string]interface{}); ok {
		linesAdded, _ = jsonInt(cost, "total_lines_added")
		linesRemoved, _ = jsonInt(cost, "total_lines_removed")
	}

	// Build render context
	ctx := &statusline.RenderContext{
		ModelName:      modelName,
		CurrentDir:     currentDir,
		ContextPercent: contextPercent,
		SessionText:    sessionText,
		UsagePercent:   usagePercent,
		LinesAdded:     linesAdded,
		LinesRemoved:   linesRemoved,
		Transcript:     transcript,
	}
	if gitInfo != nil {
		ctx.GitBranch = gitInfo.Branch
		ctx.GitUnstaged = gitInfo.Unstaged
		ctx.GitStaged = gitInfo.Staged
		ctx.GitAhead = gitInfo.Ahead
		ctx.GitBehind = gitInfo.Behind
	}

	// Load config for statusline mode
	cfg := config.LoadConfig(config.LoadConfigOptions{})
	mode := cfg.Statusline
	if mode == "" {
		mode = "full"
	}

	// Render and print
	var lines []string
	switch mode {
	case "none":
		lines = []string{""}
	case "minimal":
		lines = statusline.RenderMinimal(ctx)
	case "compact":
		lines = statusline.RenderCompact(ctx)
	default: // "full"
		lines = statusline.RenderFull(ctx)
	}

	for _, line := range lines {
		fmt.Println(line)
	}
	return nil
}

// jsonInt extracts an int from a JSON-decoded map (float64 is the default numeric type).
func jsonInt(m map[string]interface{}, key string) (int, bool) {
	v, ok := m[key]
	if !ok {
		return 0, false
	}
	switch n := v.(type) {
	case float64:
		return int(n), true
	case int:
		return n, true
	}
	return 0, false
}

// writeContextCache writes context window data to /tmp/sl-context-{sessionId}.json.
func writeContextCache(sessionID string, percent, tokens, size int, data map[string]interface{}) {
	path := filepath.Join(os.TempDir(), "sl-context-"+sessionID+".json")
	usage := map[string]interface{}{}
	if cw, ok := data["context_window"].(map[string]interface{}); ok {
		if u, ok := cw["current_usage"].(map[string]interface{}); ok {
			usage = u
		}
	}
	payload := map[string]interface{}{
		"percent":   percent,
		"tokens":    tokens,
		"size":      size,
		"usage":     usage,
		"timestamp": time.Now().UnixMilli(),
	}
	if b, err := json.Marshal(payload); err == nil {
		_ = os.WriteFile(path, b, 0644)
	}
}

// readUsageLimitsCache reads /tmp/sl-usage-limits-cache.json written by the usage-awareness hook.
// Returns (sessionText, usagePercent) where usagePercent is nil if unavailable.
func readUsageLimitsCache() (string, *int) {
	path := filepath.Join(os.TempDir(), "sl-usage-limits-cache.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", nil
	}
	var cache map[string]interface{}
	if err := json.Unmarshal(raw, &cache); err != nil {
		return "", nil
	}

	if status, _ := cache["status"].(string); status == "unavailable" {
		return "N/A", nil
	}

	cacheData, _ := cache["data"].(map[string]interface{})
	if cacheData == nil {
		return "", nil
	}
	fiveHour, _ := cacheData["five_hour"].(map[string]interface{})
	if fiveHour == nil {
		return "", nil
	}

	var usagePercent *int
	if util, ok := fiveHour["utilization"].(float64); ok {
		v := int(math.Round(util))
		usagePercent = &v
	}

	sessionText := ""
	if resetAt, ok := fiveHour["resets_at"].(string); ok && resetAt != "" {
		remaining := parseResetRemaining(resetAt)
		if remaining > 0 && remaining < 18000 {
			rh := remaining / 3600
			rm := (remaining % 3600) / 60
			sessionText = fmt.Sprintf("%dh %dm until reset", rh, rm)
		}
	}

	return sessionText, usagePercent
}

// parseResetRemaining parses an ISO 8601 timestamp and returns seconds until that time.
// Returns 0 on parse error or if the time is in the past.
func parseResetRemaining(resetAt string) int {
	t, err := time.Parse(time.RFC3339, resetAt)
	if err != nil {
		t, err = time.Parse(time.RFC3339Nano, resetAt)
		if err != nil {
			return 0
		}
	}
	remaining := int(time.Until(t).Seconds())
	if remaining < 0 {
		return 0
	}
	return remaining
}
